server/src/main/java/org/elasticsearch/index/query/IntervalQueryBuilder.java

```diff
@@ -0,0 +1,144 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.mapper.MappedFieldType;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Builder for {@link IntervalQuery}
+ */
+public class IntervalQueryBuilder extends AbstractQueryBuilder<IntervalQueryBuilder> {
+
+    public static final String NAME = "intervals";
+
+    private final String field;
+    private final IntervalsSourceProvider sourceProvider;
+
+    public IntervalQueryBuilder(String field, IntervalsSourceProvider sourceProvider) {
+        this.field = field;
+        this.sourceProvider = sourceProvider;
+    }
+
+    public IntervalQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.field = in.readString();
+        this.sourceProvider = in.readNamedWriteable(IntervalsSourceProvider.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(field);
+        out.writeNamedWriteable(sourceProvider);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(field);
+        builder.startObject();
+        sourceProvider.toXContent(builder, params);
+        printBoostAndQueryName(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    public static IntervalQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        if (parser.nextToken() != XContentParser.Token.FIELD_NAME) {
+            throw new ParsingException(parser.getTokenLocation(), "Expected [FIELD_NAME] but got [" + parser.currentToken() + "]");
+        }
+        String field = parser.currentName();
+        if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(), "Expected [START_OBJECT] but got [" + parser.currentToken() + "]");
+        }
+        String name = null;
+        float boost = 1;
+        IntervalsSourceProvider provider = null;
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                throw new ParsingException(parser.getTokenLocation(),
+                    "Expected [FIELD_NAME] but got [" + parser.currentToken() + "]");
+            }
+            switch (parser.currentName()) {
+                case "_name":
+                    parser.nextToken();
+                    name = parser.text();
+                    break;
+                case "boost":
+                    parser.nextToken();
+                    boost = parser.floatValue();
+                    break;
+                default:
+                    provider = IntervalsSourceProvider.fromXContent(parser);
+
+            }
```

Comment:
it seems that this won't fail if multiple providers are provided?

Reply:
Have updated